include/deal.II/algorithms/general_data_storage.h

```diff
@@ -0,0 +1,380 @@
+   * go out of scope. Furthermore, all scoping requirements for data referenced
+   * by this class instance will be lifted.
+   */
+  void
+  reset();
+
+  /**
+   * @name Data storage and access
+   */
+  //@{
+
+  /**
+   * Store internally a copy of the given object. The copied object is
+   * owned by this class, and is accessible via reference through the
+   * get_object_with_name() method.
+   *
+   * This function ensures that no @p entry with the given @name is
+   * already stored by this class instance.
+   */
+  template <typename Type>
+  void
+  add_unique_copy(const Type &entry, const std::string &name);
+
+  /**
+   * Store internally a copy of the given object. The copied object is
+   * owned by this class, and is accessible via reference through the
+   * get_object_with_name() method.
+   *
+   * This function does not perform any checks to ensure that the @p entry
+   * with the given @name is already stored by this class instance. If the
+   * @p name does in fact point to existing data, then this is overwritten.
+   */
+  template <typename Type>
+  void
+  add_or_overwrite_copy(const Type &entry, const std::string &name);
+
+  /**
+   * Add a reference to an already existing object. The object is not
+   * owned by this class, and the user has to guarantee that the
+   * referenced object lives longer than this class instance. The stored
+   * reference is accessible through the get_object_with_name() method.
+   *
+   * This function ensures that no @p entry with the given @name is
+   * already stored by this class instance.
+   */
+  template <typename Type>
+  void
+  add_unique_reference(Type &entry, const std::string &name);
+
+  /**
+   * Add a reference to an already existing object. The object is not
+   * owned by this class, and the user has to guarantee that the
+   * referenced object lives longer than this class instance. The stored
+   * reference is accessible through the get_object_with_name() method.
+   *
+   * This function does not perform any checks to ensure that the @p entry
+   * with the given @name is already stored by this class instance. If the
+   * @p name does in fact point to existing data, then this is overwritten.
+   */
+  template <typename Type>
+  void
+  add_or_overwrite_reference(Type &entry, const std::string &name);
+
+  /**
+   * Return a reference to the object with given name. If the object does
+   * not exist, then the input @p arguments will be used to construct an object
+   * of the given @p Type.
+   */
+  template <typename Type, typename... Args>
+  Type &
+  get_or_add_object_with_name(const std::string &name, Args &... arguments);
+
+  /**
+   * Return a reference to the object with given name. If the object does
+   * not exist, then the input @p arguments will be used to construct an object
+   * of the given @p Type.
+   */
+  template <typename Type, typename... Args>
+  Type &
+  get_or_add_object_with_name(const std::string &name, Args &&... arguments);
```

Comment:
maybe comment here what the difference to the previous function is. (namely, the rvalue references.)